client/client.go

```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
+	}
+	serviceURL, err = getAPIVersionURL(rootURL, serviceURLSuffix, versions, requestedVersion)
+	if err != nil {
+		return "", err
+	}
+	return makeURL(serviceURL, parts), nil
+}
+
+func getAPIVersionURL(rootURL *url.URL, serviceURLSuffix string, versions []apiVersionInfo, requested apiVersion) (string, error) {
+	var match string
+	for _, v := range versions {
+		if v.Version.major != requested.major {
+			continue
+		}
+		if requested.minor != -1 && v.Version.minor != requested.minor {
+			continue
+		}
+		for _, link := range v.Links {
+			if link.Rel != "self" {
+				continue
+			}
+			hrefURL, err := url.Parse(link.Href)
+			if err != nil {
+				return "", err
+			}
+			match = hrefURL.Path
+		}
+		if requested.minor != -1 {
+			break
+		}
+	}
+	if match == "" {
+		return "", fmt.Errorf("could not find matching URL")
+	}
+	versionURL := *rootURL
+	versionURL.Path = path.Join(versionURL.Path, match, serviceURLSuffix)
+	return versionURL.String(), nil
+}
+
+type apiVersion struct {
+	major int
+	minor int
+}
+
+func (v *apiVersion) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	parsed, err := parseVersion(s)
+	if err != nil {
+		return err
+	}
+	*v = parsed
+	return nil
+}
+
+func parseVersion(s string) (apiVersion, error) {
+	s = strings.TrimPrefix(s, "v")
+	parts := strings.SplitN(s, ".", 2)
+	if len(parts) == 0 || len(parts) > 2 {
+		return apiVersion{}, fmt.Errorf("invalid API version %q", s)
+	}
+	var minor int = -1
+	major, err := strconv.Atoi(parts[0])
+	if err != nil {
+		return apiVersion{}, err
+	}
+	if len(parts) == 2 {
+		var err error
+		minor, err = strconv.Atoi(parts[1])
+		if err != nil {
+			return apiVersion{}, err
+		}
+	}
+	return apiVersion{major, minor}, nil
+}
+
+type apiVersionInfo struct {
+	Version apiVersion       `json:"id"`
+	Links   []apiVersionLink `json:"links"`
+	Status  string           `json:"status"`
+}
+
+type apiVersionLink struct {
+	Href string `json:"href"`
+	Rel  string `json:"rel"`
+}
+
+type apiURLVersion struct {
+	rootURL *url.URL
+	serviceURLSuffix string
+	versions []apiVersionInfo
+}
+
+func (c *authenticatingClient) getAPIVersions(serviceCatalogURL string) (*url.URL, string, []apiVersionInfo, error) {
+	url, err := url.Parse(serviceCatalogURL)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	// Identify the version in the URL, if there is one, and record
+	// everything proceeding it. We will need to append this to the
+	// API version-specific base URL.
+	var pathParts []string
+	if url.Path != "/" {
+		pathParts = strings.Split(strings.Trim(url.Path, "/"), "/")
+		if _, err := parseVersion(pathParts[0]); err == nil {
+			pathParts = pathParts[1:]
+		}
+		url.Path = "/"
+	}
+
+	baseURL := url.String()
+
+	// make sure we haven't already received the version info
+	apiURLVersion, ok := c.apiURLVersions[baseURL]
+	if ok {
+		return apiURLVersion.rootURL, apiURLVersion.serviceURLSuffix, apiURLVersion.versions, nil
```

Comment:
instead of returning (rootURL, serviceURLSuffix, versions) why not just return the apiURLVersion struct itself